forgerock-authenticator/src/main/java/org/forgerock/android/auth/AuthenticatorManager.java

```diff
@@ -251,25 +254,81 @@ PushNotification getNotificationByMessageId(String messageId) {
+                        "device token (%s) was received.", this.pushDeviceTokenManager.getDeviceTokenId(), deviceToken);
+                throw new AuthenticatorException("The SDK was initialized with a different deviceToken. " +
+                        "Use FRAClient#updateDeviceToken method to update the device token.");
             }
         }
     }
 
+    void updateDeviceToken(String newDeviceToken, FRAListener<Void> listener) {
+        Logger.debug(TAG, "Updating FCM device token for all Push mechanisms. New token: %s", newDeviceToken);
+
+        // Get all push mechanisms
+        List<PushMechanism> pushMechanismList = getAllPushMechanisms();
+
+        // If no push mechanisms found, return success
+        if (pushMechanismList.isEmpty()) {
+            Logger.debug(TAG, "No push mechanisms found. Skipping device token update.");
+            listener.onSuccess(null);
+            return;
+        }
+
+        // Update device token for each push mechanism
+        final int totalMechanisms = pushMechanismList.size();
+        final AtomicInteger completedMechanisms = new AtomicInteger(0);
+        final List<PushMechanism> failedMechanisms = new ArrayList<>();
+        for (PushMechanism pushMechanism : pushMechanismList) {
+            pushDeviceTokenManager.updateDeviceToken(newDeviceToken, pushMechanism, new FRAListener<Void>() {
+                @Override
+                public void onSuccess(Void result) {
+                    Logger.debug(TAG, "FCM device token for mechanism %s updated successfully.", pushMechanism.getMechanismUID());
+                    if (completedMechanisms.incrementAndGet() == totalMechanisms) {
+                        if (failedMechanisms.isEmpty()) {
+                            // All mechanisms have completed successfully
+                            listener.onSuccess(null);
+                        } else {
+                            // If any mechanism failed, we report the list of failed mechanisms.
+                            listener.onException(new MechanismUpdatePushTokenException(
+                                    "Error updating FCM device token for some mechanisms.", failedMechanisms));
+                        }
+                    }
+                }
+
+                @Override
+                public void onException(Exception e) {
+                    Logger.warn(TAG, "Error updating FCM device token for mechanism %s.", pushMechanism.getMechanismUID(), e);
+                    failedMechanisms.add(pushMechanism);
+                    if (completedMechanisms.incrementAndGet() == totalMechanisms) {
+                        // All mechanisms have completed, but some failed
+                        listener.onException(new MechanismUpdatePushTokenException(
+                                "Error updating FCM device token for some mechanisms.", failedMechanisms));
+                    }
+                }
```

Comment:
Call listener.onException(e)? otherwise the caller will not receive anything.

Reply:
@witrisna, we are calling it just when all mechanisms are complete. See that we also call `listener.onException()` in `onSucess()`